build/xtask/src/sizes.rs

```diff
@@ -0,0 +1,158 @@
+                .unwrap_or_else(|| toml.stacksize.unwrap()),
+            requires: &task.requires,
+        })
+        .collect();
+    tasks.push(Task {
+        name: "kernel",
+        stacksize: toml.stacksize.unwrap(),
+        requires: &toml.kernel.requires,
+    });
+
+    let suggest = if toml.target.starts_with("thumbv7m")
+        || toml.target.starts_with("thumbv7em")
+    {
+        armv7m_suggest
+    } else if toml.target.starts_with("thumbv8m") {
+        armv8m_suggest
+    } else {
+        panic!("Unknown target {}", toml.target);
+    };
+
+    let mut suggestions = Vec::new();
+    let mut first = false; // used to print a newline between tasks
+    for task in tasks {
+        if !first {
+            first = true;
+        } else {
+            println!();
+        }
+        let mut elf_name = dist_dir.clone();
+        elf_name.push(task.name);
+        let buffer = std::fs::read(elf_name)?;
+        let elf = match Object::parse(&buffer)? {
+            Object::Elf(elf) => elf,
+            o => bail!("Invalid Object {:?}", o),
+        };
+
+        let mut memory_sizes = memories
+            .keys()
+            .map(|name| (name.as_str(), 0))
+            .collect::<IndexMap<_, _>>();
+        for phdr in &elf.program_headers {
+            if let Some(region) = output_region(phdr.p_vaddr) {
+                memory_sizes[region] += phdr.p_memsz;
+            }
+        }
+        memory_sizes["ram"] += task.stacksize as u64;
+
+        println!("{}", task.name);
+        let mut my_suggestions = Vec::new();
+        for name in memories.keys() {
+            let used = memory_sizes[name.as_str()];
+            if let Some(&size) = task.requires.get(name) {
+                let percent = used * 100 / size as u64;
+                println!(
+                    "  {:<6} {: >5} bytes ({}%)",
+                    format!("{}:", name),
+                    used,
+                    percent
+                );
+                let suggestion = suggest(used);
+                if suggestion != size as u64 {
+                    my_suggestions.push((name, size, suggestion));
+                }
+            } else {
+                assert!(used == 0);
+            }
+        }
+        if !my_suggestions.is_empty() {
+            suggestions.push((task.name, my_suggestions));
+        }
+    }
+
+    if !suggestions.is_empty() {
+        let mut savings = memories
+            .keys()
+            .map(|name| (name.as_str(), 0))
+            .collect::<IndexMap<_, _>>();
+        println!("\n========== Suggested changes ==========");
+        for (name, list) in suggestions {
+            println!("{}:", name);
+            for (mem, prev, value) in list {
+                println!(
+                    "  {:<6} {} (from {})",
+                    format!("{}:", mem),
+                    value,
+                    prev
+                );
+                savings[mem.as_str()] += prev as u64 - value;
```

Comment:
(You might consider the entry API here too)

Reply:
Done!